### PR TITLE
Put game on diet

### PR DESCRIPTION
**Pull request by teinarss**
Moving some of the code from the Game class to dedicated classes.

### CODE REVIEW

File: OpenRA.Game/Exts.cs [hunk 1]
```diff
@@ -111,12 +111,6 @@ public static bool LinesIntersect(int2 a, int2 b, int2 c, int2 d)
 			return WindingDirectionTest(c, d, a) != WindingDirectionTest(c, d, b) && WindingDirectionTest(a, b, c) != WindingDirectionTest(a, b, d);
 		}
 
-		public static bool HasModifier(this Modifiers k, Modifiers mod)
-		{
-			// PERF: Enum.HasFlag is slower and requires allocations.
```
Comment (teinarss) [inline OpenRA.Game/Exts.cs:116]: This was fixed in .net core

File: OpenRA.Game/HotkeyModifiers.cs [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+namespace OpenRA
+{
+	public class HotkeyModifiers
+	{
+		static Modifiers modifiers;
+		public static Modifiers GetModifierKeys() { return modifiers; }
+		internal static void HandleModifierKeys(Modifiers mods) { modifiers = mods; }
+
+		public static bool HasModifier(Modifiers mod)
```
Comment (pchote) [inline OpenRA.Game/HotkeyModifiers.cs:20]: Can this be removed now?

File: OpenRA.Game/Exts.cs (hunk 1 above)
Comment (pchote) [reply OpenRA.Game/Exts.cs:116]: We still have the macos compat release and people self compiling with mono, but if this is only a ~couple of percent level regression for them then IMO its worth the tradeoff to clean up the code.

File: OpenRA.Game/HotkeyModifiers.cs (hunk 2 above)
Comment (teinarss) [reply OpenRA.Game/HotkeyModifiers.cs:20]: Imo it looks better to use `HotkeyModifiers.HasModifier` than `HotkeyModifiers.GetModifierKeys().HasFlag` but its only used 5 times in the code, so im open to remove it.

File: OpenRA.Game/ExternalModSwitcher.cs [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+ */
+#endregion
+
+using System;
+using System.Diagnostics;
+using System.Linq;
+
+namespace OpenRA
+{
+	public class ExternalModSwitcher
```
Comment (pchote) [inline OpenRA.Game/ExternalModSwitcher.cs:18]: Can this be merged into the `ExternalMods` class and then called as `Game.ExternalMods.Switch()`?

File: OpenRA.Game/ExternalModSwitcher.cs [hunk 4]
```diff
@@ -0,0 +1,56 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2020 The OpenRA Developers (see AUTHORS)
```
Comment (pchote) [inline OpenRA.Game/ExternalModSwitcher.cs:3]: 2021

File: OpenRA.Game/HotkeyModifiers.cs [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+namespace OpenRA
+{
+	public class HotkeyModifiers
```
Comment (pchote) [inline OpenRA.Game/HotkeyModifiers.cs:14]: `Renderer` already handles this stuff internally and does expose some public input related state (`GetClipboardText`/`SetClipboardText`). I'm not necessarily suggesting that we expose `Renderer.HasModifier` instead, but i wonder if we're missing a more optimal abstraction that covers both.

File: OpenRA.Game/ScreenshotManager.cs [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+#endregion
+
+using System.IO;
+using OpenRA.Support;
+
+namespace OpenRA
+{
+	public class ScreenshotManager
+	{
+		public static void TakeScreenshot()
```
Comment (pchote) [inline OpenRA.Game/ScreenshotManager.cs:19]: This would probably make sense merged into `Renderer`?

File: OpenRA.Game/TextManager.cs [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+ * information, see COPYING.
+ */
+#endregion
+
+using OpenRA.Primitives;
+using OpenRA.Widgets;
+
+namespace OpenRA
+{
+	public class TextManager
```
Comment (pchote) [inline OpenRA.Game/TextManager.cs:17]: Is there any sensible owner (e.g. `ModData`) that would allow this to not be static?

**Review by pchote — COMMENTED** [review]
Thinking out loud, not necessarily review requests. I expect that some/all of these would create nasty plumbing issues due to inaccesible state.

File: OpenRA.Game/HotkeyModifiers.cs (hunk 5 above)
Comment (teinarss) [reply OpenRA.Game/HotkeyModifiers.cs:14]: Imo should we move those functions out of the Renderer class, not sure where atm.

File: OpenRA.Game/ExternalModSwitcher.cs (hunk 3 above)
Comment (teinarss) [reply OpenRA.Game/ExternalModSwitcher.cs:18]: Thought about that too, was not sure about that so started with the dedicated class.

File: OpenRA.Game/ScreenshotManager.cs [hunk 8]
```diff
@@ -0,0 +1,36 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2020 The OpenRA Developers (see AUTHORS)
```
Comment (reaperrr) [inline OpenRA.Game/ScreenshotManager.cs:3]: Likewise

File: OpenRA.Game/HotkeyModifiers.cs [hunk 9]
```diff
@@ -0,0 +1,25 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2020 The OpenRA Developers (see AUTHORS)
```
Comment (reaperrr) [inline OpenRA.Game/HotkeyModifiers.cs:3]: 2021

File: OpenRA.Game/TextManager.cs [hunk 10]
```diff
@@ -0,0 +1,51 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2020 The OpenRA Developers (see AUTHORS)
```
Comment (reaperrr) [inline OpenRA.Game/TextManager.cs:3]: Likewise

File: OpenRA.Game/ExternalModSwitcher.cs (hunk 3 above)
Comment (teinarss) [reply OpenRA.Game/ExternalModSwitcher.cs:18]: Problem is that we need to supply `launchWrapper` , we could supply it through the constructor, but we use  ExternalMods  in other places where we dont have that and also not using it.

File: OpenRA.Game/HotkeyModifiers.cs (hunk 5 above)
Comment (pchote) [reply OpenRA.Game/HotkeyModifiers.cs:14]: How about we move the `platform.CreateWindow` call into `Game.Initialize` and then pass that to the `Renderer` ctor plus a new global `Input`? We can then remove the `PumpInput` call from `Renderer.EndFrame` and have that as its own thing in `Input`.

Comment (teinarss) [reply OpenRA.Game/HotkeyModifiers.cs:14]: Dropping these changes here, will submit a new pr with changes.

File: OpenRA.Game/ScreenshotManager.cs (hunk 6 above)
Comment (teinarss) [reply OpenRA.Game/ScreenshotManager.cs:19]: Another idea is to move most of the logic from the screenshot logic code out from the renderer class and only let it return the data.

**Comment by PunkPun** [comment]
Do we want to keep this PR open? It seems we've already done many of the proposed changes and the name of the PR is fairly missleading

**Pull request closed without merging**